pinc/special_colors.inc

```diff
@@ -126,27 +123,70 @@ function echo_special_legend( $projects_where_clause)
     $day_array = array_merge($day_array, array_values(load_common_special_days()));
 
     $specs_count = count($day_array);
-    echo "<h3><a href='$code_url/tools/project_manager/show_specials.php'>$legend_text</a></h3>";
-    echo "<table class='basic'>\n";
-    $day_index = 0;
-    $table_columns = min($specs_count, 4);
-    while($day_index < $specs_count)
+    echo "<h3>$legend_text</h3>";
+    echo "<p><a href='$code_url/tools/project_manager/show_specials.php'>" . _("View all Special Days") . "</a></p>";
+
+    $num_columns = 4;
+    $items_in_columns = round(count($day_array) / $num_columns);
+    $column_index = 0;
+    $columns = [];
+    foreach($day_array as $day_entry)
     {
-        echo "<tr>\n";
-        $column_index = 0;
-        while($column_index++ < $table_columns)
+        $columns[$column_index][] = $day_entry;
+        if(count($columns[$column_index]) % $items_in_columns == 0)
         {
-            if($day_index < $specs_count)
-            {
-                $this_day = $day_array[$day_index++];
-                echo "<td style='background-color:#{$this_day['color']}'>{$this_day['display_name']}</td>\n";
-            }
-            else
-                echo "<td></td>\n";
+            $column_index += 1;
         }
-        echo "</tr>\n";
     }
-    echo "</table>\n";
+    echo "<table>\n";
+    echo "<tr>";
+    for( ; $column_index > 0; $column_index--)
+    {
+        echo "<td class='top-align'>";
+        echo "<table class='basic'>\n";
+        $column = array_shift($columns);
+        foreach($column as $day_entry)
+        {
+            list($style, $contents) = get_special_day_cell_parts($day_entry, FALSE);
+            echo "<tr>";
+            echo "<td style='$style'>$contents</td>";
```

Comment:
If special days are user controlled html safe? Are we sure the style will be ok too or should we atr safe too?

Reply:
Special Days are SA controlled. The only user input factored into the style is the color and it's reasonable to expect SAs to put in valid colors I think.